File: backend/testjam/core/sentry.py
import os
import re
import subprocess
from importlib.metadata import PackageNotFoundError, version
from typing import Any
from urllib.parse import parse_qsl, urlencode

import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
from sentry_sdk.integrations.starlette import StarletteIntegration

from testjam.core.logging import current_request_id

SCRUBBED_HEADERS = {"authorization", "cookie", "x-api-key"}
SCRUBBED_BODY_KEYS = {
    "password", "new_password", "current_password", "token", "access_token",
    "refresh_token", "api_key", "secret",
}
URL_QUERY_PATTERN = re.compile(r"^([^?]*)\?(.*)$")


def configure_sentry() -> None:
    dsn = os.environ.get("SENTRY_DSN", "").strip()
    if not dsn:
        return
    sentry_sdk.init(
        dsn=dsn,
        release=_release(),
        environment=os.environ.get("SENTRY_ENVIRONMENT", "production"),
        send_default_pii=False,
        include_local_variables=False,
        max_request_body_size="never",
        traces_sample_rate=_float_env("SENTRY_TRACES_SAMPLE_RATE", 0.0),
        profiles_sample_rate=_float_env("SENTRY_PROFILES_SAMPLE_RATE", 0.0),
        integrations=[
            StarletteIntegration(),
            FastApiIntegration(),
            SqlalchemyIntegration(),
        ],
        before_send=_before_send,
    )


def _before_send(event: dict[str, Any], _hint: dict[str, Any]) -> dict[str, Any]:
    request_id = current_request_id()
    if request_id:
        event.setdefault("tags", {})["request_id"] = request_id
    _scrub_request(event.get("request"))
    return event


def _scrub_request(request: dict[str, Any] | None) -> None:
    if not request:
        return
    headers = request.get("headers")
    if isinstance(headers, dict):
        for key in list(headers):
            if key.lower() in SCRUBBED_HEADERS:
                headers[key] = "[scrubbed]"
    data = request.get("data")
    if isinstance(data, dict):
        for key in list(data):
            if key.lower() in SCRUBBED_BODY_KEYS:
                data[key] = "[scrubbed]"
    if isinstance(request.get("query_string"), str):
        request["query_string"] = _scrub_query_string(request["query_string"])
    if isinstance(request.get("url"), str):
        request["url"] = _scrub_url(request["url"])


def _scrub_query_string(query: str) -> str:
    pairs = parse_qsl(query, keep_blank_values=True)
    if not pairs:
        return query
    cleaned = [
        (key, "[scrubbed]" if key.lower() in SCRUBBED_BODY_KEYS else value)
        for key, value in pairs
    ]
    return urlencode(cleaned)


def _scrub_url(url: str) -> str:
    match = URL_QUERY_PATTERN.match(url)
    if not match:
        return url
    base, query = match.group(1), match.group(2)
    return f"{base}?{_scrub_query_string(query)}"


def _release() -> str:
    package_version = _package_version()
    sha = _git_sha()
    if sha:
        return f"testjam-api@{package_version}+{sha}"
    return f"testjam-api@{package_version}"


def _package_version() -> str:
    try:
        return version("testjam-api")
    except PackageNotFoundError:
        return "unknown"


def _git_sha() -> str | None:
    sha = os.environ.get("GIT_SHA", "").strip()
    if sha:
        return sha[:12]
    try:
        output = subprocess.check_output(
            ["git", "rev-parse", "--short=12", "HEAD"],
            stderr=subprocess.DEVNULL,
            timeout=2,
        )
        return output.decode().strip() or None
    except (subprocess.SubprocessError, FileNotFoundError, OSError):
        return None


def _float_env(name: str, default: float) -> float:
    raw = os.environ.get(name)
    if raw is None or raw == "":
        return default
    try:
        return float(raw)
    except ValueError:
        return default

File: frontend/src/lib/sentry.js
import * as Sentry from '@sentry/react'

const SCRUB_BODY_KEYS = new Set([
  'password',
  'new_password',
  'current_password',
  'token',
  'access_token',
  'refresh_token',
  'api_key',
  'secret',
])

const SCRUB_HEADER_KEYS = new Set(['authorization', 'cookie', 'x-api-key'])

export function configureSentry() {
  const dsn = (import.meta.env.VITE_SENTRY_DSN || '').trim()
  if (!dsn) return

  Sentry.init({
    dsn,
    release: releaseTag(),
    environment: import.meta.env.VITE_SENTRY_ENVIRONMENT || import.meta.env.MODE,
    tracesSampleRate: parseFloat(import.meta.env.VITE_SENTRY_TRACES_SAMPLE_RATE || '0'),
    sendDefaultPii: false,
    beforeSend: scrubEvent,
    beforeSendTransaction: scrubEvent,
  })
}

export function scrubEvent(event) {
  const request = event?.request
  if (request) {
    request.headers = scrubMap(request.headers, SCRUB_HEADER_KEYS, (key) => key.toLowerCase())
    request.data = scrubMap(request.data, SCRUB_BODY_KEYS, (key) => key.toLowerCase())
    if (typeof request.query_string === 'string') {
      request.query_string = scrubQueryString(request.query_string)
    }
    if (typeof request.url === 'string') {
      request.url = scrubUrl(request.url)
    }
  }
  return event
}

function scrubQueryString(query) {
  const params = new URLSearchParams(query)
  let mutated = false
  for (const key of [...params.keys()]) {
    if (SCRUB_BODY_KEYS.has(key.toLowerCase())) {
      params.set(key, '[scrubbed]')
      mutated = true
    }
  }
  return mutated ? params.toString() : query
}

function scrubUrl(url) {
  const queryStart = url.indexOf('?')
  if (queryStart < 0) return url
  const base = url.slice(0, queryStart)
  const cleaned = scrubQueryString(url.slice(queryStart + 1))
  return `${base}?${cleaned}`
}

function releaseTag() {
  const explicit = (import.meta.env.VITE_APP_RELEASE || '').trim()
  if (explicit) return explicit
  const version = import.meta.env.VITE_APP_VERSION || '0.0.0'
  const sha = (import.meta.env.VITE_GIT_SHA || '').trim()
  return sha ? `testjam-frontend@${version}+${sha.slice(0, 12)}` : `testjam-frontend@${version}`
}

function scrubMap(source, sensitiveKeys, normalizeKey) {
  if (!source || typeof source !== 'object') return source
  const result = Array.isArray(source) ? [...source] : { ...source }
  for (const key of Object.keys(result)) {
    if (sensitiveKeys.has(normalizeKey(key))) {
      result[key] = '[scrubbed]'
    }
  }
  return result
}
